### PR TITLE
Clean up code review groups backend

### PR DESCRIPTION
**Pull request by bencodeorg**
This PR cleans up our Rails code after we've migrated the code review access logic to be based on code review groups, rather than the boolean per-section `code_review_enabled`. It:
- Cleans up DCDO branching conditions (such that we only use new logic)
- Removes any remaining references to the `code_review_enabled` column, such that we can drop that column.
- Fixes tests that relied on the old code review access logic, and removes tests that were testing that old logic.

## Testing story

Mostly relying on unit tests/hand testing that we've already done with the DCDO flag set to true. I did hand test that as I enabled/disabled code review via the new UI, I saw the effects of those changes as a student trying to access code review.
## Deployment strategy

Should not be merged until we've deployed code review groups via DCDO, and removed the ability to edit the code review enabled boolean in the edit/update section form.

## Follow-up work

Drop `code_review_enabled` column from `sections`.

### CODE REVIEW

**Comment by bencodeorg** [comment]
Need to fix some code review UI tests before review, please hold on review until I update.

**Comment by bencodeorg** [comment]
Note that we should keep `code_review_enabled` in the section summarize method, as we're moving around where we keep the code review enabled flag in redux here:

https://github.com/code-dot-org/code-dot-org/pull/44576/files

**Comment by bencodeorg** [comment]
Update: please review this PR despite failing UI tests, as those should be fixed once [this PR](https://github.com/code-dot-org/code-dot-org/pull/44723) is merged into this branch.

File: dashboard/test/models/ability_test.rb [hunk 1]
```diff
@@ -623,16 +623,15 @@ class AbilityTest < ActiveSupport::TestCase
     refute Ability.new(levelbuilder).can? :view_as_user, @login_required_script_level, student
   end
 
-  test 'student in same CSA code review enabled section as student seeking code review can view as peer' do
```
Comment (molly-moen) [inline dashboard/test/models/ability_test.rb:671]: should we add an additional test for students in same section but different code review groups cannot view as peer?

**Review by molly-moen — APPROVED** [review]
Nice cleanup! One idea for an additional test, otherwise looks good!

File: dashboard/test/models/ability_test.rb (hunk 1 above)
Comment (bencodeorg) [reply dashboard/test/models/ability_test.rb:671]: mmm good idea, will add.

**Comment by bencodeorg** [comment]
@davidsbailey, looking for a bit of expertise on seeding -- I needed to seed the CSA course to get some UI tests to work (we only show this "Code Review Group" management UI on the teacher dashboard for sections assigned a CSA script.)

We only seed a couple of courses currently (a couple versions of CSP?) Is adding CSA particularly bad practice? I don't know what the additional seeding time is for adding an additional course.

I really only need the course, and a single script from it, but I think seeding the course will fail unless I seed all of the associated scripts?

**Comment by davidsbailey** [comment]
> @davidsbailey, looking for a bit of expertise on seeding -- I needed to seed the CSA course to get some UI tests to work (we only show this "Code Review Group" management UI on the teacher dashboard for sections assigned a CSA script.)
> 
> We only seed a couple of courses currently (a couple versions of CSP?) Is adding CSA particularly bad practice? I don't know what the additional seeding time is for adding an additional course.
> 
> I really only need the course, and a single script from it, but I think seeding the course will fail unless I seed all of the associated scripts?

hi @bencodeorg , our preferred strategy for UI testing is to run our tests against test-only scripts. Following our existing naming convention, for you that would mean creating a script like allthejavalabthings or allthecsathings and enabling the features you want to test there. Will that work?

to create a new script, you'll either need to create it on levelbuilder, or (probably better) run your local server in levelbuilder_mode and create the new script locally via /s/new .

**Comment by bencodeorg** [comment]
> > @davidsbailey, looking for a bit of expertise on seeding -- I needed to seed the CSA course to get some UI tests to work (we only show this "Code Review Group" management UI on the teacher dashboard for sections assigned a CSA script.)
> > We only seed a couple of courses currently (a couple versions of CSP?) Is adding CSA particularly bad practice? I don't know what the additional seeding time is for adding an additional course.
> > I really only need the course, and a single script from it, but I think seeding the course will fail unless I seed all of the associated scripts?
> 
> hi @bencodeorg , our preferred strategy for UI testing is to run our tests against test-only scripts. Following our existing naming convention, for you that would mean creating a script like allthejavalabthings or allthecsathings and enabling the features you want to test there. Will that work?
> 
> to create a new script, you'll either need to create it on levelbuilder, or (probably better) run your local server in levelbuilder_mode and create the new script locally via /s/new 

Testing this feature requires having a section that is assigned to a CSA course, or one of its scripts. I currently do that by creating a section assigned to CSA via the section create form, which I don't think would work if I created a dummy script?

A lighter weight alternative to seeding the whole course and its scripts might be to create a section assigned to a CSA script via HTTP request, so we'd only have to seed one script. I was just able to hack something like this by hard coding the correct script ID in the section creation step here:

https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/ui/features/step_definitions/section_management_steps.rb#L3-L8

https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/api/v1/sections_controller.rb#L31-L44

That said, I don't know how I could get the right script ID consistently across environments?

**Comment by davidsbailey** [comment]
> Testing this feature requires having a section that is assigned to a CSA course, or one of its scripts.

This sounds like a problem -- we shouldn't have features which depend on the names of specific courses, rather we should have features which we enable or disable for each course. Can you point me to the code which controls this? a possible compromise would be to base this on the "core course" or "curriculum umbrella", which we would have the ability to set on a test script.

**Comment by bencodeorg** [comment]
> > Testing this feature requires having a section that is assigned to a CSA course, or one of its scripts.
> 
> This sounds like a problem -- we shouldn't have features which depend on the names of specific courses, rather we should have features which we enable or disable for each course. Can you point me to the code which controls this? a possible compromise would be to base this on the "core course" or "curriculum umbrella", which we would have the ability to set on a test script.

Sure -- we hide/show the UI based on this boolean:

https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/sections/section.rb#L458-L460

Didn't realize we could set umbrella on a test script, that could work! How would you suggest creating the section then?

**Comment by davidsbailey** [comment]
> Didn't realize we could set umbrella on a test script, that could work! How would you suggest creating the section then?

ok, great! to create a section in a UI test, you would either go through the UI steps needed to create a section, or you would send an ajax request to an API in test_controller.rb and have that create a section for you. for example, you can search for `assign_script_as_student` to see a codepath that may already be doing something similar.

**Comment by bencodeorg** [comment]
> > Didn't realize we could set umbrella on a test script, that could work! How would you suggest creating the section then?
> 
> ok, great! to create a section in a UI test, you would either go through the UI steps needed to create a section, or you would send an ajax request to an API in test_controller.rb and have that create a section for you. for example, you can search for `assign_script_as_student` to see a codepath that may already be doing something similar.

@davidsbailey will test UI scripts show up in the list of scripts in the section create UI? The existing section create steps seem more geared towards selecting a multi-script course (eg, CSP) vs. selecting a specific script. 

Just confirming, is your concern about seeding the CSA scripts that it's overkill to seed 9 scripts for a handful of tests? Or more that the test will regress if/when the `csa-pilot` course is deleted? Or something else/a combination of the two?

**Comment by davidsbailey** [comment]
> will test UI scripts show up in the list of scripts in the section create UI? The existing section create steps seem more geared towards selecting a multi-script course (eg, CSP) vs. selecting a specific script.

if you mark them as launched / stable, they should show up in the assignment selector dropdown. if that doesn't work or is too much work or too slow, you could use test_controller.rb to create the section assignment for you. Please feel free to ping me tomorrow if you would like help with creating the test script and/or generally getting this working.

> Just confirming, is your concern about seeding the CSA scripts that it's overkill to seed 9 scripts for a handful of tests? Or more that the test will regress if/when the `csa-pilot` course is deleted? Or something else/a combination of the two?

it is a combination of the two. at present, adding 9 scripts unfortunately has the potential to slow down drone runs, and we've been working to trim the list of existing UI_TEST_SCRIPTS down. also, tests which depend on content owned by curriculum writers can be fragile in other situations, such as when new versions of CSA are launched each year.

**Comment by bencodeorg** [comment]
Note for future reference if the test script needs to be updated at any point (these steps were made with advice from Dave):
- I created a script containing a single (empty) lesson under the CSA "curriculum umbrella" via our normal levelbuilder UI.
- This created a couple of script_json files -- I manually moved the script config file to the test directory (dashboard/test/ui/config/scripts_json/ui-test-csa-family-script.script_json), and deleted any other files that were generated.
- There were diffs created by adding this script in scripts.en.yml, a lot of which were adding blank fields to already existing scripts. I found the section in scripts.en.yml that related to this new script, copied it, reset scripts.en.yml to its state before any changes were made, and manually copied the new config to the bottom of the file.

**Comment by bencodeorg** [comment]
@davidsbailey @molly-moen could you take a quick look at this commit (excuse the rogue puts, deleted those in a [subsequent commit](https://github.com/code-dot-org/code-dot-org/pull/44320/commits/a9f3f530143b86090429250b588e8ff0f896f07f)):

https://github.com/code-dot-org/code-dot-org/pull/44320/commits/396df2a7f8fb2ff90c2b15b149ff90500b472854

There are a bunch of new commits here, but they're mostly repeatedly resolving merge conflicts.

**Review by davidsbailey — APPROVED** [review]
looks great Ben, nice job getting this working with ajax!

**Review by molly-moen — APPROVED** [review]
LGTM!

**Pull request merged**